### ISSUE TITLE
Feedback on the persist() algorithm.

### ISSUE THREAD

**Issue by michael-nordman** [whatwg/storage#26]
Not necessarily bugs, but a couple things I noticed.

If |permission| is "prompt" on entry, there is no way the promise can resolve without setting |permission| to either "grant" or "deny". If a user agent should wish to implement a "Not Now" dismissal of the prompt, the only spec compliant implementation would be to not resolve the promise. I think a resolution to false in that case would be more helpful so logic in the page would not be expecting user input that will never arrive.

Is the ASSERT added between 4 and 5 below expected to always hold true? The steps below may allow the promise to resolve to "true" despite the user having just chosen "deny". That seems wrong.

```
1. Let origin be settingsObject’s origin.

2. Let permission be permission for origin.

3. If permission is "prompt", then ask the user whether turning origin’s site storageunit’s box
   into a persistent box is acceptable. If it is, then set permission to "granted", and "denied" otherwise.

4. Let persisted be true if origin’s site storage unit’s box is a persistent box, and false otherwise.

  if (permission == "denied") ASSERT(!persisted);

5. Queue a task to run these subsubsteps:

    1., Set permission for origin to permission.

    2. If persisted is false and permission is "granted", set persisted to true and set origin’s
       site storage unit’s box’s mode to "persistent".

    3. Resolve promise with persisted.
```

**Comment by annevk** [comment]
> I think a resolution to false in that case would be more helpful so logic in the page would not be expecting user input that will never arrive.

So this case also came up for the Notifications API and the tentative conclusion there is to resolve with "prompt". I don't know what the general approach for dismissal is in the Permissions API. @jyasskin, @mounirlamouri?

In Storage I tried an approach where `persist()` would mainly be about attempting to persist and tell you the result, but if the dismissal case is important you might want to know about it, so then `persist()` should resolve with a tuple, or simply with the permission state.

(The assert should hold true.)

**Comment by mounirlamouri** [comment]
The Permissions API has the same approach: `request()` actually returns the new permission state so if nothing changed (ie. the user dismissed), it will return "prompt". That also solves the cases where a permission is requested when the user previously granted/denied it, in which case it will return "granted"/"denied" respectively.

**Comment by michael-nordman** [comment]
> (The assert should hold true.)

I mentioned on another issue that handling concurrent calls to persist() has interesting implementation details. Imagine two instances of this algo running with two separate prompts on screen, the user says "grant" to the first prompt and runs to completion, then a moment later the user says "deny" to the second.

I think the assert would not hold true in such a world.

To reduce the risk of compat issues, it'd be good to address the potential inconsistency between the 'permission' value and the 'persist' value. I'd vote to alter the spec to explicitly set 'persist' to a value consistent with the user input.

Maybe there's a good reason the algo doesn't currently do that?

**Comment by annevk** [comment]
Basically, the assumption is that you can only go from non-persistent to persistent. If the user allows in one tab, I think it should cause the dialog to be dismissed in the other tab and resolve with true (or maybe "granted" since we want to allow resolving with "prompt" too I guess).

**Comment by michael-nordman** [comment]
That behavior sounds reasonable but its not fully reflected in the algo, there is no mention of waiting on an existing prompt for that origin to complete. From an impl point of view, with these prompts appearing, as 'banners' scoped to a tab, having a user interaction with one of the prompts affect the state of another prompt is extra complexity, i think the path of least resistance would be to allow transitioning from persistent to non-persistent in this hypothetical case, but that would be non-compliant as there's no way the algo here could produce that result.

Maybe add some non-normative words, equivalent to the ASSERT, or some other clarifying statement of the assumptions.

**Comment by annevk** [comment]
Yeah sounds reasonable. Would you be okay with waiting for other tabs to complete?

**Comment by michael-nordman** [comment]
> Would you be okay with waiting for other tabs to complete?

I'm not sure I understand what your asking?

I think the algo could be specified without a particular user experience in mind so its agnostic enough for browser vendors to provide different UX yet still be spec-compliant. I'm not sure what chrome does today for overlapping prompts in two or more tabs, but our UX designers should be able to change their minds about that tomorrow.

I think altering the algo to recognize that  origin.permission and origin.persist may be modified while the prompt is up would allow for the most flexibility in UX and implementations.

**Comment by annevk** [comment]
I decided that making `persist()` do the opposite of what it promises would be too weird and therefore encouraged user agents through the standard to not enable such a scenario.

I hope that's acceptable.
